### PR TITLE
Add FetchReference() for Repository Target

### LINKED ISSUE #122: Add Function to Fetch Manifest Using Reference
- Currently, v2 requires a manifest descriptor to be resolved first before calling `Fetch` to get the artifact manifest
- Propose to add a function that will take in a reference and return the artifact manifest

### PR DESCRIPTION
**Pull request by Wwwsylvia**
Closes #122

### CODE REVIEW

File: registry/repository.go [hunk 1]
```diff
@@ -65,6 +65,7 @@ type BlobStore interface {
 	content.Storage
 	content.Deleter
 	content.Resolver
+	TagFetcher
```
Comment (shizhMSFT) [inline registry/repository.go:68]: Can we also make `Repository` implement `TagFetcher`, which defaults for manifests?

File: registry/remote/repository.go [hunk 2]
```diff
@@ -601,6 +608,56 @@ func (s *blobStore) Resolve(ctx context.Context, reference string) (ocispec.Desc
...
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := s.repo.client().Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if err != nil {
```
Comment (shizhMSFT) [inline registry/remote/repository.go:636]: which `err` are you referring to?

File: registry/repository.go [hunk 3]
```diff
@@ -75,6 +76,14 @@ type TagPusher interface {
 	PushTag(ctx context.Context, expected ocispec.Descriptor, content io.Reader, reference string) error
 }
 
+// TagFetcher provides advanced fetch with the tag service.
+type TagFetcher interface {
+	// FetchTag fetches the content identified by a reference.
+	// It is equivalent to call `Resolve()` and then `Fetch()`
+	// but more efficient or equal.
+	FetchTag(ctx context.Context, reference string) (io.ReadCloser, error)
```
Comment (shizhMSFT) [inline registry/repository.go:84]: It is better to return
```go
(io.ReadCloser, ocispec.Descriptor, error)
```

File: registry/remote/repository.go [hunk 4]
```diff
@@ -739,6 +796,58 @@ func (s *manifestStore) Resolve(ctx context.Context, reference string) (ocispec.
...
+		return nil, err
+	}
+	req.Header.Set("Accept", manifestAcceptHeader(s.repo.ManifestMediaTypes))
+
+	resp, err := s.repo.client().Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if err != nil {
```
Comment (shizhMSFT) [inline registry/remote/repository.go:821]: same here

**Review by shizhMSFT — CHANGES_REQUESTED** [review]

File: registry/remote/repository.go (hunk 2 above)
Comment (Wwwsylvia) [reply registry/remote/repository.go:636]: Oops!

File: registry/repository.go (hunk 3 above)
Comment (Wwwsylvia) [reply registry/repository.go:84]: Good idea!

File: registry/remote/repository.go [hunk 5]
```diff
@@ -739,6 +796,58 @@ func (s *manifestStore) Resolve(ctx context.Context, reference string) (ocispec.
 	}, nil
 }
 
+// FetchTag fetches the content identified by a reference.
+// It is equivalent to call `Resolve()` and then `Fetch()`
```
Comment (sajayantony) [inline registry/remote/repository.go:800]: This should not be equivalent of resolve and fetch. They are 2 API requests where are this should be only one GET request.

**Review by sajayantony — CHANGES_REQUESTED** [review]
Does this have both casss of GET by digest and by tag name?

File: registry/remote/repository.go (hunk 5 above)
Comment (shizhMSFT) [reply registry/remote/repository.go:800]: It is equivalent in terms of functionality. We need to mention that it is more efficient as it has only 1 GET request in the documentation.

Comment (sajayantony) [reply registry/remote/repository.go:800]: I would agree but would like to keep the docs for the function without ambiguity. It is clearly 2 APIs and they have different behaviors depending on intermediaries.  

Maybe edit to keep it concise and have more details in the example. 
`//FetchTag fetches the content identified by the reference. The reference may be a tag or digest.`

Comment (sajayantony) [reply registry/remote/repository.go:800]: On now thinking more - its sounds incorrect to call this method `FetchTag` if we are passing a digest. 
We should consider some other name like `FetchByReference(string reference....)`. 
Or am I missing something here?

Comment (Wwwsylvia) [reply registry/remote/repository.go]: Agreed. I've rename it to `FetchReference` and made the doc concise.

**Comment by sajayantony** [comment]
The renames look really good. The fact that we are standardizing on ref seems much more consistent.  Do you have a test case for this format 

`docker image pull k8s.gcr.io/pause-amd64:some-made-up-tag@sha256:59eec8837a4d942cc19a52b8c09ea75121acc38114a2c68b98983ce9356b8610`

**Comment by Wwwsylvia** [comment]
> The renames look really good. The fact that we are standardizing on ref seems much more consistent. Do you have a test case for this format
> 
> `docker image pull k8s.gcr.io/pause-amd64:some-made-up-tag@sha256:59eec8837a4d942cc19a52b8c09ea75121acc38114a2c68b98983ce9356b8610`

Test cases added.

**Comment by Wwwsylvia** [comment]
@shizhMSFT @sajayantony Could you please review the updates?

File: registry/remote/repository.go [hunk 6]
```diff
@@ -583,17 +595,72 @@ func (s *blobStore) Resolve(ctx context.Context, reference string) (ocispec.Desc
 	default:
 		return ocispec.Descriptor{}, errutil.ParseErrorResponse(resp)
 	}
+
+	return s.generateDescriptor(resp, refDigest)
```
Comment (shizhMSFT) [inline registry/remote/repository.go:599]: nit: let put this line to line 592

File: registry/remote/repository.go [hunk 7]
```diff
@@ -583,17 +595,72 @@ func (s *blobStore) Resolve(ctx context.Context, reference string) (ocispec.Desc
...
+	default:
+		return ocispec.Descriptor{}, nil, errutil.ParseErrorResponse(resp)
+	}
+
+	desc, err = s.generateDescriptor(resp, refDigest)
+	if err != nil {
+		return ocispec.Descriptor{}, nil, err
+	}
+
+	return desc, resp.Body, nil
```
Comment (shizhMSFT) [inline registry/remote/repository.go:645]: After checking RFC 7233, it is possible to have `httputil.ReadSeekCloser` for this method, where the request should have a range header `Range: bytes=0-`.

Comment (Wwwsylvia) [reply registry/remote/repository.go]: Nice

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Review by sajayantony — APPROVED** [review]

**Comment by sajayantony** [comment]
@qweeah will you add an example for this one?

**Pull request merged**